dali/kernels/transpose/transpose_gpu_impl.cuh

```diff
@@ -0,0 +1,308 @@
+  T *__restrict__ out;
+  const T *__restrict__ in;
+  uint64_t in_strides[kMaxNDim];
+  // the output strides are permuted and it's this permutation that defines the whole operation
+  uint64_t out_strides[kMaxNDim];
+  fast_div<uint64_t> shape[kMaxNDim];
+
+  uint32_t tiles_y;
+  uint32_t tiles_per_block;
+  uint32_t total_tiles;
+  int ndim;
+  int lanes;
+
+  fast_div<uint32_t> tiles_x;
+  fast_div<uint32_t> tiles_per_slice;
+};
+
+namespace transpose_shared {
+  extern __shared__ uint8_t shared_tmp[];
+}  // namespace transpose_shared
+
+
+template <int ndim, typename T>
+__device__ void TransposeTiledStatic(TiledTransposeDesc<T> desc) {
+  unsigned start_tile = blockIdx.x * desc.tiles_per_block;
+  unsigned end_tile = min(desc.total_tiles, start_tile + desc.tiles_per_block);
+  if (start_tile >= end_tile)
+    return;
+
+  T (*tmp)[kTileSize][kTileSize+1] =
+      reinterpret_cast<T (*)[kTileSize][kTileSize+1]>(transpose_shared::shared_tmp);
+
+  int lanes = desc.lanes;
+
+  unsigned tile_in_slice = start_tile;
+  uint64_t fused_slice = ndim > 2
+    ? div_mod(tile_in_slice, start_tile, desc.tiles_per_slice)
+    : 0;
+
+  uint64_t pos[ndim];  // NOLINT
+
+  for (int d = ndim - 3; d >= 0; d--) {
+    fused_slice = div_mod(pos[d], fused_slice, desc.shape[d]);
+  }
+
+  unsigned tile_x, tile_y;
+  tile_y = div_mod(tile_x, tile_in_slice, desc.tiles_x);
+  pos[ndim - 1] = tile_x * kTileSize;
+  pos[ndim - 2] = tile_y * kTileSize;
+
+  T *out = desc.out;
+  const T *in = desc.in;
+
+  for (uint64_t tile = start_tile;;) {
+    uint64_t in_ofs = 0, out_ofs = 0;
+    #pragma unroll
+    for (int d = 0; d < ndim - 2; d++) {
+      in_ofs  += desc.in_strides[d] * pos[d];
+      out_ofs += desc.out_strides[d] * pos[d];
+    }
+
+    int64_t in_x  = pos[ndim-1] + threadIdx.x;
+    int64_t in_y  = pos[ndim-2] + threadIdx.y;
+    int64_t out_x = pos[ndim-1] + threadIdx.y;
+    int64_t out_y = pos[ndim-2] + threadIdx.x;
+
+    in_ofs  += desc.in_strides[ndim-2]  * in_y  + desc.in_strides[ndim-1]  * in_x;
+    out_ofs += desc.out_strides[ndim-2] * out_y + desc.out_strides[ndim-1] * out_x;
+
+    int tile_w = min(static_cast<uint64_t>(kTileSize), desc.shape[ndim-1] - pos[ndim-1]);
+    int tile_h = min(static_cast<uint64_t>(kTileSize), desc.shape[ndim-2] - pos[ndim-2]);
+    if (threadIdx.x < tile_w) {
+      for (int ty = threadIdx.y, dy = 0; ty < tile_h; ty += blockDim.y, dy += blockDim.y) {
+        #pragma unroll(4)
+        for (int lane = 0; lane < lanes; lane++) {
+          tmp[lane][ty][threadIdx.x] = __ldg(&in[in_ofs + desc.in_strides[ndim-2]*dy + lane]);
+        }
+      }
+    }
+    __syncthreads();
+
+    if (threadIdx.x < tile_h) {
+      for (int ty = threadIdx.y, dy = 0; ty < tile_w; ty += blockDim.y, dy += blockDim.y) {
+        #pragma unroll(4)
+        for (int lane = 0; lane < lanes; lane++)
+          out[out_ofs + desc.out_strides[ndim-1]*dy + lane] = tmp[lane][threadIdx.x][ty];
+      }
+    }
+
+    if (++tile >= end_tile)
```

Comment:
the algorithm would be nicer to read if you put this in the for loop:
```for (uint64_t tile = start_tile; tile < end_tile; tile++) {```

and here you just did:
```
if (tile < end_tile - 1) {  // avoid advancing the offset and __syncthreads in the last tile
  ...
}
```

Reply:
This would cause the condition to be evaluated twice - and does putting just the increment in the loop really help?